### DISCUSSION TITLE
Support for Automatic Indexing

### DISCUSSION THREAD

**Discussion by Baroshem** [nuxt-modules/algolia#36] [Ideas]
## Intro

The Algolia module for Nuxt now has all basic functionality that is related to search and recommend (and it also provides a support for Vue Instantsearch). Now we could focus on integrating it with other services like Content Management System. 

I have recently released an article for Storyblok where I was explaining how to build a Nuxt 3 application with Algolia and Storyblok (including automatic indexing of indices using Storyblok publish hooks) -> https://www.storyblok.com/tp/using-storyblok-algolia-in-nuxt-3

Similar article was also created for Contentful -> https://www.contentful.com/developers/docs/tutorials/general/enhancing-search-experience-with-algolia/

This article explains it quite well, however it requires some manual work from the user in order to make it work, so I thought about automating it so that user would only have to define some configuration option for algolia module in nuxt.config.ts and the indexing will be done automatically (Not fully automatically because it will still require a serverless function or triggering an endpoint manually but you get the idea ;) ).

## Code

Option 1:

```ts
// nuxt.config.js

{
    buildModules: [
        "@nuxtjs/algolia",
    ],
    algolia: {
        apiKey: "<YOUR_SEARCH_API_KEY>",
        applicationId: "<YOUR_APPLICATION_ID>",
        indexer: {
            cms: {
                name: 'storyblok | contentful',
                options: {
                  accessToken: '123',
                  // For Storyblok
                  indexName: 'storyblok',
                  // For Contentful
                  spaceId: '213daw21'
                }
             }
        }
    }
}
```

Option 2:

```ts
// nuxt.config.js

{
    buildModules: [
        "@nuxtjs/algolia",
    ],
    algolia: {
        apiKey: "<YOUR_SEARCH_API_KEY>",
        applicationId: "<YOUR_APPLICATION_ID>",
        indexer: {
            storyblok: {
              accessToken: '123',
              indexName: 'storyblok'
            },
            contentful: {
              accessToken: '123',
              spaceId: '123dawdaw1'
            }
        }
    }
}
``` 

Indexer option is optional, while selecting a storyblok or contentful (or in the future more providers) will be mandatory when indexer option will be enabled. By having all these env variables, we could easily trigger a StoryblokAlgoliaIndexer from the package I have created or a ContentfulAlgoliaIndexer in the future.

Under the hood, this new feature would create an appriopriate endpoint/serverMiddleware to handle this kind of functionality for the user.

## Summary

As a part of this we would also have to create a docs on how to use both of the CMSes with option of indexing.

**Comment by Baroshem** [comment]
AVuengers assemble!

Let me know what you think about it :)

@danielroe @pi0 @alexjoverm @alvarosabu @timbenniks @Atinux @yassilah

**Comment by chuckmeyer** [comment]
I'm all for reducing friction for the developer experience! 

I recently did a deep dive on adding Algolia to a Strapi powered blog (Next front end, not Nuxt -- sorry!). There are definitely commonalities between these ingestion patterns and it would be super cool to somehow normalize them. But I wonder if the application framework is the right place for this? 

These would be provisioned as serverless functions to compliment the application?

**Reply by Baroshem** [reply]
Hey @chuckmeyer 

Yeah, it can be achieved by having a serverless function. However, I would like to provide an optional support for people who would prefer to have a server rendered application with endpoint for triggering indexing so that if they dont want to have additional external serverless function, they could just utilize the same application but just with an additional endpoint.

If they want to use the serverless function instead, they can always not use the optional indexer and build one by themselves. The main idea behind this RFC is to give the possibility to the users so that it will be done by them (maybe not the perfect way as serverless is a better place but if they dont want to invest more time, they could just utilize the built in feature of the module).

What do you think?
